README
NAME
    Parallel::Scoreboard - a scoreboard for monitoring status of many
    workers

SYNOPSIS
      use Parallel::Scoreboard;

      my $scoreboard = Parallel::Scoreboard->new(
          base_dir => '/tmp/my_scoreboard'
      ...

      # in each worker process
      $scoreboard->update('this is my current status');

      # to read status of all worker processes
      my $stats = $scoreboard->read_all();
      for my $pid (sort { $a <=> $b } keys %$stats) {
          print "status for pid:$pid is: ", $stats->{$pid}, "\n";
      }

DESCRIPTION
    Parallel::Scoreboard is a pure-perl implementation of a process
    scoreboard. By using the module it is easy to create a monitor for many
    worker process, like the status module of the Apache HTTP server.

    Unlike other similar modules, Parallel::Scoreboard is easy to use and
    has no limitation on the format or the length of the statuses to be
    stored. Any arbitrary data (like JSON or frozen perl object) can be
    saved by the worker processes as their status and read from the manager
    process.

METHODS
  new(%args)
    instantiation. Recognizes the following paramaters. The parameters can
    be read using the read-only accessors with the same name.

   base_dir => $base_dir
    the directory name in which the scoreboard files will be stored. The
    directory will be created if it does not exist already. Mandatory
    parameter.

   worker_id => sub { ... }
    a subref that returns the id of the worker (if omitted, the module uses
    $$ (process id) to distinguish the workers)

  update($status)
    saves the status of the process

  read_all()
    reads the status of all worker processes that are alive and that have
    called update() more than once. Returned value is a hashref with process
    ids as keys and the statuses of each processes as corresponding values.

  cleanup()
    remove obsolete status files found in base_dir. The files are normally
    removed upon the termination of worker process, however they might be
    left unremoved if the worker process was killed for some reason. The
    detection and removal of the obsolete status files is performed by
    read_all() as well.

SEE ALSO
    IPC::ScoreBoard Proc::Scoreboard

AUTHOR
    Kazuho Oku <kazuhooku gmail.com>

LICENSE
    This program is free software, you can redistribute it and/or modify it
    under the same terms as Perl 5.10.



Changes
Revision history for Perl module Parallel::Scoreboard

0.07 2015-01-26
 - do not include MYMETA.* in the distribution

0.06 2015-01-26
 - create workspace directory by using mkpath instead of mkdir (tolid)
 - fixed a warning that could occur during global destruction when an attribute in the scoreboard object was destroyed before the object itself (Dave Rolsky)

0.05 2014-04-09
 - when cleaning up obsolete scoreboard files, ignore errors when another process cleaned up the file at the same time (Karen Etheridge)

0.04 2013-08-19
 - fix seek error (thanks to hirose31)
 - support non-numeric ids (thanks to hirose31)

0.03 2011-01-24
 - added callback "worker_id" that can be used for distiguishing workers other than using pids

0.02 2010-07-04
 - fix incompatibility with Solaris (LOCK_EX not possible on read-only file handles)
 - scripts may now call fork(2) after calling Parallel::Scoreboard::update()

0.01 2010-07-02
 - initial release


lib/Parallel/Scoreboard/PSGI/App.pm
package Parallel::Scoreboard::PSGI::App;

use Class::Accessor::Lite;
use HTML::Entities;
use Parallel::Scoreboard;

use strict;
use warnings;

Class::Accessor::Lite->mk_accessors(qw(scoreboard));

sub new {
    my $klass = shift;
    my %args = @_;
    die "mandatory parameter:scoreboard is missing"
        unless $args{scoreboard};
    # build object
    my $self = bless {
        %args,
    }, $klass;
    return $self;
}

sub to_app {
    my $self = shift;
    return sub {
        my $status = $self->scoreboard->read_all;
        return [
            200,
            [ 'Content-Type' => 'text/html; charset=utf-8' ],
            [
                sprintf(
                    << 'EOT',
<html>
  <head>
    <title>Status of %s</title>
    <meta http-equiv="content-type" content="text/html; charset=utf-8" />
  </head>
  <body>
    <h1>Status of %s</h1>
    <hr />
    <table>
      <tr><th>PID</th><th>Status</th></tr>
EOT
                    encode_entities($self->scoreboard->base_dir),
                    encode_entities($self->scoreboard->base_dir),
                ),
                (map {
                    sprintf(
                        '<tr><td>%s</td><td>%s</td>',
                        encode_entities($_),
                        encode_entities($status->{$_}),
                    )
                } sort { $a <=> $b } keys %$status),
                sprintf(
                    << 'EOT',
    </table>
    <hr />
    Generated by <a href="http://search.cpan.org/dist/Parallel-Scoreboard/">Parallel::Scoreboard</a> %s
  </body>
</html>
EOT
                    encode_entities($Parallel::Scoreboard::VERSION),
                ),
            ],
        ];
    };
}

1;


lib/Parallel/Scoreboard/PSGI/App/JSON.pm
package Parallel::Scoreboard::PSGI::App::JSON;

use Class::Accessor::Lite;
use HTML::Entities;
use JSON;
use Parallel::Scoreboard;

use strict;
use warnings;

Class::Accessor::Lite->mk_accessors(qw(scoreboard));

sub new {
    my $klass = shift;
    my %args = @_;
    die "mandatory parameter:scoreboard is missing"
        unless $args{scoreboard};
    # build object
    my $self = bless {
        %args,
    }, $klass;
    return $self;
}

sub to_app {
    my $self = shift;
    return sub {
        my $status = $self->scoreboard->read_all;
        for my $pid (%$status) {
            $status->{pid} = decode_json($status->{pid})
                if $status->{$pid} =~ /^[\{\[]/;
        }
        return [
            200,
            [ 'Content-Type' => 'application/json; charset=utf-8' ],
            [ encode_json($status) ],
        ];
    };
}

1;
__END__

=head1 NAME

Parallel::Scoreboard::PSGI::App::JSON - a simple PSGI app for monitoring the output of Parallel::Scoreboard in JSON format

=head1 SYNOPSIS

  use Parallel::Scoreboard;
  use Parallel::Scoreboard::PSGI::App;
  
  my $scoreboard = Parallel::Scoreboard->new(
      base_dir => '/tmp/my_scorebooard',
  );
  # return psgi app
  Parallel::Scoreboard::PSGI::App->new(
      scoreboard => $scoreboard,
  )->to_app;

=head1 SEE ALSO

L<Parallel::Scoreboard>
L<Parallel::Scoreboard::PSGI::App::JSON>
L<PSGI>

=head1 AUTHOR

Kazuho Oku E<lt>kazuhooku gmail.comE<gt>

=head1 LICENSE

This program is free software, you can redistribute it and/or modify it under the same terms as Perl 5.10.

=cut


lib/Parallel/Scoreboard.pm
package Parallel::Scoreboard;

use strict;
use warnings;

use Digest::MD5 qw(md5);
use Fcntl qw(:flock);
use IO::Handle;
use POSIX qw(:fcntl_h);
use File::Path;

our $VERSION = '0.07';

use Class::Accessor::Lite (
    ro => [ qw(base_dir worker_id) ],
);

sub new {
    my $klass = shift;
    my %args = @_;
    die "mandatory parameter:base_dir is missing"
        unless $args{base_dir};
    # create base_dir if necessary
    if (! -e $args{base_dir}) {
	mkpath $args{base_dir}
            or die "failed to create directory:$args{base_dir}:$!";
    }
    # build object
    my $self = bless {
        worker_id => sub { $$ },
        %args,
    }, $klass;
    # remove my status file, just in case
    unlink $self->_build_filename();
    
    return $self;
}

sub DESTROY {
    my $self = shift;
    # if file is open, close and unlink
    if ($self->{fh}) {
        close $self->{fh};
        # during global destruction we may already have lost this
	unlink $self->_build_filename() if ($self->{base_dir});
    }
}

sub update {
    my ($self, $status) = @_;
    # open file at the first invocation (tmpfn => lock => rename)
    my $id = $self->worker_id->();
    if ($self->{fh} && $self->{id_for_fh} ne $id) {
        # fork? close but do not unlock
        close $self->{fh};
        undef $self->{fh};
    }
    unless ($self->{fh}) {
        my $fn = $self->_build_filename();
        open my $fh, '>', "$fn.tmp"
            or die "failed to open file:$fn.tmp:$!";
        autoflush $fh 1;
        flock $fh, LOCK_EX
            or die "failed to flock file:$fn.tmp:$!";
        rename "$fn.tmp", $fn
            or die "failed to rename file:$fn.tmp to $fn:$!";
        $self->{fh} = $fh;
        $self->{id_for_fh} = $id;
    }
    # write to file with size of the status and its checksum
    seek $self->{fh}, 0, SEEK_SET
        or die "seek failed:$!";
    print {$self->{fh}} (
        md5($status),
        pack("N", length $status),
        $status,
    );
}

sub read_all {
    my $self = shift;
    my %ret;
    $self->_for_all(
        sub {
            my ($id, $fh) = @_;
            # detect collision using md5
            for (1..10) {
                seek $fh, 0, SEEK_SET
                    or die "seek failed:$!";
                my $data = do { local $/; join '', <$fh> };
                # silently ignore if data is too short
                return if length($data) < 16 + 4;
                # parse input
                my $md5 = substr($data, 0, 16);
                my $size = unpack("N", substr($data, 16, 4));
                my $status = substr($data, 20, $size);
                # compare md5 to detect collision
                next
                    if md5($status) ne $md5;
                # have read correct data, save and return
                $ret{$id} = $status;
                return;
            }
            # failed to read data in 10 consecutive attempts, bug?
            warn "failed to read status of id:$id, skipping";
        }
    );
    \%ret;
}

sub cleanup {
    my $self = shift;
    $self->_for_all(sub {});
}

sub _for_all {
    my ($self, $cb) = @_;
    my @files = glob "$self->{base_dir}/status_*";
    for my $fn (@files) {
        # obtain id from filename (or else ignore)
        $fn =~ m|/status_(.*)$|
            or next;
        my $id = $1;
        # ignore files removed after glob but before open
        open my $fh, '+<', $fn
            or next;
        # check if the file is still opened by the owner process using flock
        if ($id ne $self->worker_id->() && flock $fh, LOCK_EX | LOCK_NB) {
            # the owner has died, remove status file
            close $fh;
            unlink $fn
                or
                    not $!{ENOENT} and
                    warn "failed to remove an obsolete scoreboard file:$fn:$!";
            next;
        }
        # invoke
        $cb->($id, $fh);
        # close
        close $fh;
    }
}

sub _build_filename {
    my $self = shift;
    return "$self->{base_dir}/status_" . $self->worker_id->();
}

1;
__END__

=head1 NAME

Parallel::Scoreboard - a scoreboard for monitoring status of many workers

=head1 SYNOPSIS

  use Parallel::Scoreboard;

  my $scoreboard = Parallel::Scoreboard->new(
      base_dir => '/tmp/my_scoreboard'
  ...

  # in each worker process
  $scoreboard->update('this is my current status');

  # to read status of all worker processes
  my $stats = $scoreboard->read_all();
  for my $pid (sort { $a <=> $b } keys %$stats) {
      print "status for pid:$pid is: ", $stats->{$pid}, "\n";
  }

=head1 DESCRIPTION

Parallel::Scoreboard is a pure-perl implementation of a process scoreboard.  By using the module it is easy to create a monitor for many worker process, like the status module of the Apache HTTP server.

Unlike other similar modules, Parallel::Scoreboard is easy to use and has no limitation on the format or the length of the statuses to be stored.  Any arbitrary data (like JSON or frozen perl object) can be saved by the worker processes as their status and read from the manager process.

=head1 METHODS

=head2 new(%args)

instantiation.  Recognizes the following paramaters.  The parameters can be read using the read-only accessors with the same name.

=head3 base_dir => $base_dir

the directory name in which the scoreboard files will be stored.  The directory will be created if it does not exist already.  Mandatory parameter.

=head3 worker_id => sub { ... }

a subref that returns the id of the worker (if omitted, the module uses $$ (process id) to distinguish between the workers)

=head2 update($status)

saves the status of the process

=head2 read_all()

reads the status of all worker processes that are alive and that have called update() more than once.  Returned value is a hashref with process ids as keys and the statuses of each processes as corresponding values.

=head2 cleanup()

remove obsolete status files found in base_dir.  The files are normally removed upon the termination of worker process, however they might be left unremoved if the worker process was killed for some reason.  The detection and removal of the obsolete status files is performed by read_all() as well.

=head1 SEE ALSO

L<IPC::ScoreBoard>
L<Proc::Scoreboard>

=head1 AUTHOR

Kazuho Oku E<lt>kazuhooku gmail.comE<gt>

=head1 LICENSE

This program is free software, you can redistribute it and/or modify it under the same terms as Perl 5.10.

=cut
